### DISCUSSION TITLE
Using ompython

### DISCUSSION THREAD

**Discussion by ap766** [OpenModelica/OpenModelica#11419] [Q&A] [answered]
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/998c0c0a-ec8f-42b8-a4c4-97e4ef186e81)

It says loadFile Error: Error: Failed to insert class TwoConnectedTanks within NonInteractingTanks;

 the available classes were:.........

I have followed the documentation.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/786becde-6739-410b-8883-7ceb0b8f2935)

here's some of the files.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/356a6500-1a07-4cbf-b01e-5859b2a1ef0b)
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/9109ea2e-d5b9-4a09-8ce8-c21fd51dd85f)
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/f6b70cfb-4e28-4863-8222-748e13aae6d7)

please help.

**Comment by arun3688** [comment]
@ap766 The model is inside a package so you should use the constructor like the following way, 
```

>>> from OMPython import ModelicaSystem
>>> mod =ModelicaSystem("package.mo", "NonInteractingTanks.TwoConnectedTanks")
```

It would be good if you can clearly state what you want to achieve and if you can provide your models I can help you little bit more. I can see that you are trying to use OMEdit but the explanation was not clear there, But based on some assumptions we are trying to help you

**Comment by ap766** [comment]
thanks a lot!

So actually i want to use OMPython because I wish to give the value of the input variables and get the output variables .I wish to do it in a way other than doing it in openmodelica directly because I am creating a python GUI that would be taking the inputs and creating the XML and .mat file based on the outputs that I get for the input values . And then I would be opening only the .mat file in openmodelica  such that the variables browser also shows the new output values calculated .

This is how my model is 
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/2aaf32d1-de2f-4766-b10a-a6afc4798093)

1)Clicking on tank1, gives me these input variables.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/45db04a1-462d-4feb-be6c-564f7ba47261)

2)Clicking on tank2, gives these input variables.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/9dd174f7-b4f7-43ce-bc3a-90cc53d02c73)

3)Now when I simulate the twoconnectedtanks.mo by clicking on simulate - variables browser / output variables  are shown and we select the ones we wish to plot.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/e0d5c3b5-44a6-44dd-b2ca-fb46b0b09aea)

So in this particular case i didn't give any input of mine in 1) and 2) so it gave me the default in 3) , but suppose i do give input in 1) or 2) , values in 3) (variables browser/output variables) would be calculated according. 

And i wish to achieve this in a way other than doing it in openmodelica directly ,

till now 1) i was using a command -`C:\Users\lenovo\Documents>TwoConnectedTanks.exe -override=startTime=0,stopTime=5,tank1.A=4,tank1.h=7


stdout            | warning | Start or stop time was overwritten, but no new integrator step size was provided.
|                 | info    | | Re-calculating step size for 500 intervals.
|                 | |       | | Add `stepSize=<value>` to `-override=` or override file to silence this warning.
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_SUCCESS       | info    | The simulation finished successfully.`

But i realised i was just overriding the values directly nd not giving input values , but directly overriding the output values.

2)i thought of modifying init.xml but again even here i would be modifying only the output variables directly and not the input variables.

So wish to find a way that i can get output variables based on input variables.So thought maybe OMPython is the only way

Actually the connection still couldn't be made and shows an error.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/be439ea9-a77a-4615-a9e9-9bbb96750a94)

Here's the drive link if you wish to see the code for the files - https://drive.google.com/drive/folders/1wK2Qy9-_6bOgXtoTLM-pfdWgHRjtBuOP?usp=share_link . Also please ignore the (1) here , it exists only because i have uploaded it in drive.

**Comment by ap766** [comment]
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/3a716101-9e14-483f-a322-8131cef2c2f5)
just another screenshot if it helps

**Comment by arun3688** [comment, marked as answer]
@ap766  Hello i got your files and found that your package structured is wrong and that is why you got the above error, I have fixed your package but i also found another error that your model is not balanced, I cannot help much on this part since i am not a modeller.see below

```
>>> from OMPython import ModelicaSystem
>>> mod = ModelicaSystem("package.mo", "NonInteractingTanks.TwoConnectedTanks")
Error: Too few equations, under-determined system. The model has 8 equation(s) and 9 variable(s).
[c:/OPENMODELICAGIT/NonInteractingTanks/FlowConnect.mo:4:3-4:25:writable] Warning: Variable tank2.flowConnect.potentialVariable does not have any remaining equation to be solved in.
  The original equations were:
  Equation 1: tank1.flowConnect.potentialVariable = tank2.flowConnect.potentialVariable, which needs to solve for tank1.flowConnect.potentialVariable
```



See the attached zip file. Next time you can put all your source file as a zip and then post it here as that would be much easier, Look for the python script `simulate.py` see the below zip file


[NonInteractingTanks.zip](https://github.com/OpenModelica/OpenModelica/files/13063296/NonInteractingTanks.zip)

**Reply by ap766** [reply]
Thank you so much!

**Comment by arun3688** [comment]
Have you fixed your models, If you have fixed your model, then

getOutputs() should give you the list of Outputs, you can use simulate()
API, to simulate the model, otherwise please fix your model and send me th
zip file.


regards
arun

**Reply by ap766** [reply]
So, actually that was a sample model , I am supposed to implement it on this.

So I realised that setparameters actually is for setting the input variables and getparameters shows the input variables.But the output variables are in getquantities I suppose? but the changes do not reflect here nor in the XML file. (eg v - I did setparameters to 10, but getquanties shows 100 only , v is still supposed to show the same value and obviously it can be done in other ways like modifying the XML file , but I need to get all the output variables (which are calculated and hence I wouldn't know the values)

On using getOutputs I get an empty dictionary as the output even after I use simulate .

**Reply by ap766** [reply]
If i do not enter any inputs , it can take the default input variables , but if i do enter even a single input variable , i need that value to be considered and get output variable accordingly so i just need to know where this output variable is stored.

**Reply by arun3688** [reply]
@ap766 you have to use `setInputs()` API,  please go through the documentation carefully https://openmodelica.org/doc/OpenModelicaUsersGuide/1.21/ompython.html#standard-set-methods, First start with a simple example and see everything works in the way you wanted and then you should proceed with complicated models, I suppose this is your first time working with OpenModelica

**Reply by ap766** [reply]
okay thanks a lot.yes its my first time.

Actually documentation wasn't very clear because I couldn't find any variable called cAi and Ti in the bouncingballs model file
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/22e64553-ac07-4163-9e17-fbadc0dc1150)

**Reply by arun3688** [reply]
@ap766 the documentation shows an example, You have to adjust the API to your examples

**Reply by ap766** [reply]
I request you ,if possible , to help me find out why mod.setInputs(["batch_Rectifier.V=10"]) doesn't work... it changes neither mat nor XML files , and obviously isn't giving me the output results.

Followed the right syntax.
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/5c53a9f9-7c1a-42f6-8b07-88a60eaf5474)


[Simle_Batch_Distillation.zip](https://github.com/OpenModelica/OpenModelica/files/13071505/Simle_Batch_Distillation.zip)
Here's the zip file

**Reply by arun3688** [reply]
@ap766 Because `batch_Rectifier.V` is not an input and you would have clearly got an error message that is not an input something like below

```
>>> mod.setInputs(["batch_Rectifier.V=10.0"])
batch_Rectifier.V !is not an input
```
Working Solution where you can see the output, First you should differentiate between normal variable and output variable, A variable is considered output only if it has the definition `output Real x;` in the model, According to the model provided there is no output variable and you can see it from the API `getOutputs()` see below 

```
>>> from OMPython import ModelicaSystem
>>> mod = ModelicaSystem("package.mo", "Simle_Batch_Distillation.Test")
>>> mod.getInputs()
{}
>>> mod.getOutputs()
{}
>>> mod.getParameters()
{'batch_Rectifier.HB0': '300.0', 'batch_Rectifier.HD': '10.0', 'batch_Rectifier.HN': '1.0', 'batch_Rectifier.V': '100.0', 'batch_Rectifier.XB0[1]': '0.3', 'batch_Rectifier.XB0[2]': '0.3', 'batch_Rectifier.XB0[3]': '0.4', 'batch_Rectifier.alpha[1]': '9.0', 'batch_Rectifier.alpha[2]': '3.0', 'batch_Rectifier.alpha[3]': '1.0', 'batch_Rectifier.NOC': '3', 'batch_Rectifier.NT': '40'}
>>> mod.setParameters(["batch_Rectifier.V=10.0"])
>>> mod.getParameters("batch_Rectifier.V")
['10.0']
>>> mod.simulate()
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_SUCCESS       | info    | The simulation finished successfully.
>>> mod.getSolutions("batch_Rectifier.V")
array([[10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10.,
        10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10.,
        10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10., 10.,
        10., 10., 10., 10., 10., 10., 10., 10.]])

>>> mod.resultfile
'C:/Users/arupa54/AppData/Local/Temp/tmpc4qrk8il/Simle_Batch_Distillation.Test_res.mat'
```
Now if you plot the above result file in the plot window you can see the new value 10, I suggest you to read the documentation carefully as all the examples are available there

**Reply by ap766** [reply]
thank you so much 

![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/2333543e-c551-4431-b67c-4ca4149fdf29)
I considered this to be the input.......
....................................................................................................................................................................................................................

![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/b1a39cf2-4799-40d5-b3ed-a9f091d67a2e)
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/658ab280-e734-4bf3-b236-6b8212aa9c6f)
And these variables to be the output............ is there any way to get these variables that are calculated exclusively based on input , using ompython

**Reply by arun3688** [reply]
@ap766 I guess you are not reading the documentation, you can use the `getSolutions()` API as i mentioned in the previous solutions

```
>>> mod.getSolutions()  /// will list you all the variables 
>>> mod.getSolutions("L") 
```

something like above

**Reply by ap766** [reply]
I am sorry ! thanks a lot

**Comment by ap766** [comment]
Just another thing i had followed the documentation , so I am trying to override a value that I have -
RigorousBatchDistillation.exe -override=startTime=0,stopTime=5,NOC=7.This is the description(1) I changed some things because it is not reflecting in the .mat file .But at same time if I am doing it for something like RigorousBatchDistillation.exe -override=startTime=0,stopTime=5,PC=6 it reflects on my file .(Image 2)

1)
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/bfcf4443-1e81-4fbf-8567-f6a9704afc41)
2)
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/412b9108-4b66-4eb2-9960-5f4c3ac15e07)
............So could you please tell me why it doesn't reflect on the plot...is it something like because its a integer and only real values can be changed like this?

**Comment by ap766** [comment]
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/e941fb13-c899-46db-833e-b642a44f3835)


the noc description was also like this...I changed things to see if they will reflect on the plot...but still didnt

**Reply by AnHeuermann** [reply]
You are not allowed to change every variable of the simulation at each point in the simulation.
Some variables can only be changed before compiling the model, some only at initialization, some at specific points during simulations and others never.
It depends on the causality and variability.
E.g for your example the causality is `calculatedParameter` and the variability is `parameter`. An equation is added to the initial system where the value is calculated, so it can't be changed directly. You would need to change the variables that are used to calculate this parameter.

For Modelica I don't know of a good description of this. For FMI 2.0.4 this is standardized in section `2.2.7 Definition of Model Variables (ModelVariables)`.
